src/JustSaying/AwsTools/QueueCreation/SnsWriteConfiguration.cs

```diff
@@ -11,6 +10,6 @@ public class SnsWriteConfiguration
         /// Extension point enabling custom error handling on a per notification basis, including ability handle raised exceptions.
         /// </summary>
         /// <returns>Boolean indicating whether the exception has been handled</returns>
-        public Func<Exception, Message, bool> HandleException { get; set; }
+        public Func<Exception, object, bool> HandleException { get; set; }
```

Comment:
Maybe we'd need to bite the bullet and introduce the generics you mentioned? Otherwise, using the error handler is going to be quite bleurgh from a caller having to cast/pattern-match the type all the time.

Reply:
Yeah, this is the bit that persuaded me we need to do better. As it stands there are islands of type information, we can fix that by having more generic apis at the top level and flowing it down. A big change with this approach would be `IMessagePublisher<T>` and reworking the builder APIs (yet again).

Or we do the message envelope thing, which is looking more appealing. It would contain `Id` & `Timestamp` (perhaps) and a `TMessage` property, but other than that, is really just a way of allowing people a clean way to "pick out" their message types by defining a concrete envelope, then they get the nice typed guide rails they'd come to expect.

The fiddly bit for us with this approach is then splitting out the properties on (de)serializing, but shouldn't be too tricky.

Reply:
Quite the conundrum! I'm not such a fan of the idea of `IMessagePublisher<T>`, although if we did choose to go down that route, we could probably easily have some sort of `IAnyMessagePublisher`.

I think it's worth thinking about what use there was in having this typed as `Message` in the first place. What sort of error handling code would we like to be able to write? If you want to write some generic error handling code then you can't write that in terms of any _specific_ domain message type anyway. In that case you'd rather want to write a try-catch in your handler for that _specific_ message.

Reply:
That's my thoughts too @shaynevanasperen, I think a message envelope concept doesn't solve this problem, it acts as a potential way to separate the non-domain stuff from the transporty stuff.

When the base class was `Message`, you were either:
- using `Id`, `Tenant` etc... on `Message`
- ignoring the message and doing some generic error reporting (I think this is the primary use case)
- upcasting to the known inheriting types

You could still do those last two with a message of type `object`, it just looks bad.

Reply:
Oh, I just realised that this is for handling SNS publishing errors!  I thought it was for errors in implementations of `IHandlerAsync<T>`. I've had a look around the code now and I see that the equivalent of this on the other side (SQS) is in [SqsReadConfiguration](https://github.com/justeat/JustSaying/blob/master/src/JustSaying/AwsTools/QueueCreation/SqsReadConfiguration.cs#L27). Over there, we work exclusively with the `Amazon.SQS.Model.Message` type (with no strong typing of our domain message).

So why not have symmetry with that approach and just do the equivalent here? That would mean using `Amazon.SimpleNotificationService.Model.PublishRequest` making the signature `public Func<Exception, PublishRequest, bool> HandleException { get; set; }`. The `PublishRequest` object has the message as a serialized string, so the error logging can easily use that. This abstraction seems like it should remain at the raw transport level, and if consumers want more control they can always wrap the `IMessagePublisher` this library returns with some decorator that adds custom error logging and possibly even things like a retry mechanism etc.

While looking around at the [code which uses this configuration delegate](https://github.com/justeat/JustSaying/blob/master/src/JustSaying/AwsTools/MessageHandling/SnsTopicBase.cs#L58-L70), I noticed something else that we might want to do there. The error handling there only handles exceptions but won't call our error handler if the `response.HttpStatusCode` is not 200. I haven't checked this though (maybe the `Client.PublishAsync` call does throw for non-success status codes?).